docs/adr/0028-remplacer-orm-bookshelfjs-par-query-builder-knexjs.md

```diff
@@ -0,0 +1,50 @@
+Les avantages les plus notables mis en avant par les éditeurs/utilisateurs de telles librairies sont séduisants :
+- Apporter de l'abstraction vis à vis des détails techniques d'interaction avec la BDD. La promesse des ORMs est que le code produit sera compatible quelle que soit le système de base de données relationnel (MySQL, PostgreSQL, SQLite, etc...). 
+- Alléger la charge mentale du développeur en ne lui imposant plus une connaissance fine du langage SQL.
+- Facilement déclarer et requêter sur des objets du domaine lesquels sont de parfois de façon complexe stockée en BDD (multi-tables, relations étrangères, etc...).
+- Améliorer la lisibilité des interactions avec la BDD en faisant en sorte d'intéragir avec des entités métiers.
+
+Les avantages sont très intéressants sur le papier, mais en réalité cela s'avère être un peu plus compliqué
+
+### Les difficultés rencontrées aujourd'hui
+
+Aujourd'hui, la plateforme Pix doit gérer tous les jours un nombre important de requêtes utilisateurs, et donc doit gérer un nombre tout aussi important d'intéractions avec la BDD.
+A la fois du point de vue des performances mais aussi du quotidien du développeur, BookshelfJS présente de plus en plus d'inconvénients, qui sont en fait connus des ORMs et qui constituent leurs limites.
+
+- Afin de tenir leurs promesses, les ORMs sont en réalité des libraires complexes avec une interface riche. Il s'avère finalement que la montée en compétence sur l'usage de l'outil (et la maintenance du code associé) n'est pas si évidente.
+- Les ORMs, à partir du code produit en utilisant leur librairie, génèrent automatiquement les requêtes SQL associées. Pour des requêtes simples sur des objets simples de type CRUD, le SQL généré est plutôt efficace. Malheureusement, dès lors que l'on souhaite réaliser des opérations un peu plus complexes, cela devient vite une lutte avec la librairie. De plus, souvent, celle-ci va générer un SQL pas toujours performant.
+```js
+// Exemple inspiré de code existant de récupération 
+// d'un assessment-result avec ses competence-marks
+const assessmentResultBookshelf = await BookshelfAssessmentResult
+  .where({ id: 123 })
+  .fetch({ withRelated: ['competenceMarks'] });
+```
+```sql
+# SQL généré
+SELECT "assessment-results".* FROM "assessment-results" WHERE "id" = $1 LIMIT $2; (bindings : {123, 1})
+SELECT DISTINCT "competence-marks".* FROM "competence-marks" WHERE "competence-marks"."assessmentResultId" IN ($1); (bindings : {123})
+```
+On constate que :
+1. Deux requêtes sont générées alors que cela pourrait être réalisé en une seule (à noter cela dit que BookshelfJS prend soin d'exécuter ces requêtes dans la même transaction).
+2. La syntaxe est destinée à fonctionner pour tellement de cas différents qu'elle est trop générique et peut ralentir la requête (ici le DISTINCT n'est pas nécessaire par exemple).
+
```

Comment:
3- On ne parle plus du soucis de conso mémoire : les étapes successives de serialization en objet bookshelf puis du domaine, une ligne plus tard et qui prennent beaucoup de place en mémoire pour pas grand chose ?

Reply:
Et les objets Bookshelf sont particulièrement lourds parce qu'ils contiennent de la logique de suivi des modifications, etc.

Reply:
oui oui c'est en development in progress ! ca arrive vous inquiétez pas j'ai les chiffres